# Introduction
This is a dataset scraped from eBay Kleinanzeigen, a classifieds section of the German eBay website, and then uploaded to Kaggle. 50,000 data points have been sampled from the set for this exercise.

The data will be cleaned and analyzed for each used car listing. The data dictionary provided with data is as follows:

- dateCrawled - When this ad was first crawled. All field-values are taken from this date.
- name - Name of the car.
- seller - Whether the seller is private or a dealer.
- offerType - The type of listing
- price - The price on the ad to sell the car.
- abtest - Whether the listing is included in an A/B test.
- vehicleType - The vehicle Type.
- yearOfRegistration - The year in which the car was first registered.
- gearbox - The transmission type.
- powerPS - The power of the car in PS.
- model - The car model name.
- kilometer - How many kilometers the car has driven.
- monthOfRegistration - The month in which the car was first registered.
- fuelType - What type of fuel the car uses.
- brand - The brand of the car.
- notRepairedDamage - If the car has a damage which is not yet repaired.
- dateCreated - The date on which the eBay listing was created.
- nrOfPictures - The number of pictures in the ad.
- postalCode - The postal code for the location of the vehicle.
- lastSeenOnline - When the crawler saw this ad last online.



In [1]:
import pandas as pd 
import numpy as np 

In [2]:
autos = pd.read_csv("autos.csv", encoding="Latin-1")
autos.info()
print("\n")
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

Based on the data above, there are 50,000 entries in the data frame and 20 columns to help describe them. Some of the columns include dateCrawled, name, yearOfRegistration, odometer, and price.

Some column have null values, but all have relatively low null values overall, at around less than 20%. 

# Cleaning the columns
The columns use camelcase rather than snakecase, which can be more difficult to work with. Some of the columns could be more descriptive than they are as well. We will change these issues first. 

In [3]:
autos.columns



Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [4]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_pictures', 'postal_code',
       'last_seen']

autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  num_pictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

# Exploring the data
In order to further clean the data, we need to explore the dataset. We will be looking for columns with text values that are all or mostly the same, numeric data stored as text that can be converted, anything else to note that could use cleaning. 

In [5]:
autos.describe(include="all")

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-30 17:37:35  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  num_pictures   postal_code            last_seen  
count                 50000       50000.0  50000.000000                50000  
unique                   76           NaN           NaN                39481  
top     2016-04-03 00:00:00           NaN           NaN  2016-04-07 06:17:27  
freq                   1946           NaN           NaN                    8  
mean                    NaN           0.0  50813.627300                  NaN  
std                     NaN           0.0  25779.747957                  NaN  
min                     NaN           0.0   1067.000000                  NaN  
25%                     NaN           0.0  30451.000000                  NaN  
50%                     NaN           0.0  49577.000000                  NaN  
75%                     NaN           0.0  71540.000000                  NaN  
max                     NaN           0.0  99998.000000                  NaN

At a glance it appears that "seller" and "offer_type" columns only have 2 unique values. Based on the column description, it seems these should be higher. 

"num_pictures" has NaN or 0.0 values for everything except "count". We will explore this further. 

In [6]:
autos["num_pictures"].value_counts()

0    50000
Name: num_pictures, dtype: int64

In [7]:
autos["num_pictures"].head()





0    0
1    0
2    0
3    0
4    0
Name: num_pictures, dtype: int64

# Cleaning the data
It appears that "num_pictures" is essentially empty, with 0 for every value. Due to this we will delete the column along with the "seller" and "offer_type" columns, since they also do not hold accurate or unique data. 

In [8]:
autos = autos.drop(["num_pictures", "seller", "offer_type"], axis=1)

In [9]:
autos.columns

Index(['date_crawled', 'name', 'price', 'ab_test', 'vehicle_type',
       'registration_year', 'gearbox', 'power_ps', 'model', 'odometer',
       'registration_month', 'fuel_type', 'brand', 'unrepaired_damage',
       'ad_created', 'postal_code', 'last_seen'],
      dtype='object')

Above we are checking to ensure the columns were properly removed. 

The "price" and "odometer" columns have numeric values stored as text. This is due to "$" , "," and "km" from initial observation. We will remove these characters  and convert the column to a numeric dtype. Then remame the "odometer" column to "odometer_km" to ensure accuracy. 



In [10]:
autos["price"] = (autos["price"].str.replace("$","")
.str.replace(",","").astype(int)
                 )
autos["price"].head()

0    5000
1    8500
2    8990
3    4350
4    1350
Name: price, dtype: int64

In [11]:
autos["odometer"] = (autos["odometer"].str.replace("km","").str.replace(",","").astype(int)
                    )
autos.rename({"odometer": "odometer_km"}, axis=1, inplace=True)

autos["odometer_km"].head()

0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer_km, dtype: int64

Both "price" and "odometer_km" appear to have been converted to int types successfully and cleaned. 

Next, we'll continue cleaning the data by looking for and removing outliers. We will start with the "odometer_km" and "price" columns first. 

In [12]:
print(autos["odometer_km"].unique().shape)
print("\n")
print(autos["odometer_km"].describe())
print("\n")
print(autos["odometer_km"].value_counts(ascending=False))

(13,)


count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64


150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: odometer_km, dtype: int64


At a glance, there don't seem to be any glaring outliers. However, it seems every car owner/dealer had to choose from a set mileage, as there are no exact mileage numbers listed. This leads to the question does the large number of vehicles with 150,000 miles actually have than number of miles, or do some vehicles have much higher than than and there wasn't an option? For our purposes, we cannot change this. But that is something to keep in mind if we choose to use eBay's information agian in the future for cars. 

Next we'll take a look at the "price" column.

In [13]:
print(autos["price"].unique().shape)
print("\n")
print(autos["price"].describe())
print("\n")
print(autos["price"].value_counts().sort_index(ascending=False).head(15))
print("\n")
print(autos["price"].value_counts().sort_index(ascending=False).tail(15))


(2357,)


count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64


99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
Name: price, dtype: int64


18       1
17       3
15       2
14       1
13       2
12       3
11       2
10       7
9        1
8        1
5        2
3        1
2        3
1      156
0     1421
Name: price, dtype: int64


Prices in the "price" column vary widely. There are a lare number of statistically significant outliers. We will remove the outliers based on the standard deviation we found above. 

# Removing outliers based on standard deviation

Since there are 50,000 vehicles to look through, it's best to let math do the work for us. Any sales that were more or less than 3 standard deviations away from the mean will be dropped from the "price" column. 

The mean is 9840.04376 and the SD is 481104.38050019875. This means our limits are 1,443,313.14 for the upper limit and -1433473.09774059625 as the lower limit. Of course a negative number won't be used in eBay car sales, so we would typically leave the number at 0. 

In [14]:
autos = autos[autos["price"].between(0,1443313.14)]
autos["price"].describe()



count    4.999100e+04
mean     5.831664e+03
std      1.427337e+04
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.300000e+06
Name: price, dtype: float64

As you can see, a small number of data has been removed from "price", but the statistics have significantly changed due to the outliers being removed. The mean has gone from $9840 to $5,831. The standard deviation has went from $481,104 to $5,831. 

# Date columns

There are 5 columns with date information:

- "date_crawled": added by the crawler
- "last_seen": added by the crawler
- "ad_created": from the website
- "registration_month": from the website
- "registration_year": from the website

The first three are all stored as string values, so we'll need to convert them into int values. The last two columns can be explored as is. 


In [15]:
autos[['date_crawled', 'ad_created', 'last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

The first 10 characters of each value represent what we need. We can take that and leave the rest.

In [16]:
(autos["date_crawled"].str[:10].value_counts(normalize=True,dropna=False).sort_index())

2016-03-05    0.025385
2016-03-06    0.013943
2016-03-07    0.035966
2016-03-08    0.033266
2016-03-09    0.033206
2016-03-10    0.032126
2016-03-11    0.032486
2016-03-12    0.036767
2016-03-13    0.015563
2016-03-14    0.036627
2016-03-15    0.033986
2016-03-16    0.029505
2016-03-17    0.031526
2016-03-18    0.013062
2016-03-19    0.034906
2016-03-20    0.037827
2016-03-21    0.037507
2016-03-22    0.032926
2016-03-23    0.032386
2016-03-24    0.029105
2016-03-25    0.031746
2016-03-26    0.032486
2016-03-27    0.031046
2016-03-28    0.034846
2016-03-29    0.034166
2016-03-30    0.033626
2016-03-31    0.031906
2016-04-01    0.033806
2016-04-02    0.035406
2016-04-03    0.038687
2016-04-04    0.036507
2016-04-05    0.013102
2016-04-06    0.003181
2016-04-07    0.001420
Name: date_crawled, dtype: float64

These were all crawled between March and April 2016. They were crawled at a fairly consistent rate. 

In [17]:
(autos["ad_created"].str[:10].value_counts(normalize=True,dropna=False).sort_index())

2015-06-11    0.000020
2015-08-10    0.000020
2015-09-09    0.000020
2015-11-10    0.000020
2015-12-05    0.000020
2015-12-30    0.000020
2016-01-03    0.000020
2016-01-07    0.000020
2016-01-10    0.000040
2016-01-13    0.000020
2016-01-14    0.000020
2016-01-16    0.000020
2016-01-22    0.000020
2016-01-27    0.000060
2016-01-29    0.000020
2016-02-01    0.000020
2016-02-02    0.000040
2016-02-05    0.000040
2016-02-07    0.000020
2016-02-08    0.000020
2016-02-09    0.000040
2016-02-11    0.000020
2016-02-12    0.000060
2016-02-14    0.000040
2016-02-16    0.000020
2016-02-17    0.000020
2016-02-18    0.000040
2016-02-19    0.000060
2016-02-20    0.000040
2016-02-21    0.000060
                ...   
2016-03-09    0.033226
2016-03-10    0.031866
2016-03-11    0.032786
2016-03-12    0.036607
2016-03-13    0.016923
2016-03-14    0.035226
2016-03-15    0.033746
2016-03-16    0.030005
2016-03-17    0.031206
2016-03-18    0.013722
2016-03-19    0.033846
2016-03-20    0.037867
2016-03-21 

Most of these adds were created in the year 2016, but some were created in the second half of 2015. The created consistency seems to be slow in the first half but picks up in the second half. It's nothing that causes statistical concern. 

In [18]:
(autos["last_seen"].str[:10].value_counts(normalize=True,dropna=False).sort_index())

2016-03-05    0.001080
2016-03-06    0.004421
2016-03-07    0.005361
2016-03-08    0.007581
2016-03-09    0.009842
2016-03-10    0.010762
2016-03-11    0.012522
2016-03-12    0.023804
2016-03-13    0.008982
2016-03-14    0.012802
2016-03-15    0.015883
2016-03-16    0.016443
2016-03-17    0.027925
2016-03-18    0.007421
2016-03-19    0.015743
2016-03-20    0.020704
2016-03-21    0.020724
2016-03-22    0.021584
2016-03-23    0.018583
2016-03-24    0.019564
2016-03-25    0.019203
2016-03-26    0.016963
2016-03-27    0.016023
2016-03-28    0.020864
2016-03-29    0.022344
2016-03-30    0.024844
2016-03-31    0.023824
2016-04-01    0.023104
2016-04-02    0.024904
2016-04-03    0.025365
2016-04-04    0.024624
2016-04-05    0.124282
2016-04-06    0.220980
2016-04-07    0.130944
Name: last_seen, dtype: float64

"last_seen" is typically the date the add was removed due to being sold. All the dates are in 2016. The removal remains consistent up until the last 3 days of the data. This likely wasn't a huge increase in removal of cars, but a removal due to the crawling ending. We'll revisit this if it becomes an issue. 

# Registration year

In the "registration_year" column, the minimum value is 1000 and the max value is 9999. These are inaccurate. To make these more accurate, we'll remove all values that are not between 1903 (when the first license plate became mandatory in some places) and 2016 (when this data was taken).

At a quick glance, we can tell that the numbers being removed do not represent a statistically significant number. It will be safe to remove them.

In [19]:
autos["registration_year"].value_counts(normalize=True).sort_index()

1000    0.000020
1001    0.000020
1111    0.000020
1500    0.000020
1800    0.000040
1910    0.000180
1927    0.000020
1929    0.000020
1931    0.000020
1934    0.000040
1937    0.000080
1938    0.000020
1939    0.000020
1941    0.000040
1943    0.000020
1948    0.000020
1950    0.000060
1951    0.000040
1952    0.000020
1953    0.000020
1954    0.000040
1955    0.000040
1956    0.000100
1957    0.000040
1958    0.000080
1959    0.000140
1960    0.000660
1961    0.000120
1962    0.000080
1963    0.000180
          ...   
2001    0.054050
2002    0.050669
2003    0.054550
2004    0.054750
2005    0.060311
2006    0.054150
2007    0.046088
2008    0.044628
2009    0.041968
2010    0.031946
2011    0.032686
2012    0.026465
2013    0.016123
2014    0.013302
2015    0.007981
2016    0.026325
2017    0.029045
2018    0.009822
2019    0.000060
2800    0.000020
4100    0.000020
4500    0.000020
4800    0.000020
5000    0.000080
5911    0.000020
6200    0.000020
8888    0.000020
9000    0.0000

In [20]:
autos = autos[autos["registration_year"].between(1903,2016)]
autos["registration_year"].value_counts(normalize=True).sort_index()

1910    0.000187
1927    0.000021
1929    0.000021
1931    0.000021
1934    0.000042
1937    0.000083
1938    0.000021
1939    0.000021
1941    0.000042
1943    0.000021
1948    0.000021
1950    0.000062
1951    0.000042
1952    0.000021
1953    0.000021
1954    0.000042
1955    0.000042
1956    0.000104
1957    0.000042
1958    0.000083
1959    0.000146
1960    0.000687
1961    0.000125
1962    0.000083
1963    0.000187
1964    0.000250
1965    0.000354
1966    0.000458
1967    0.000562
1968    0.000541
          ...   
1987    0.001562
1988    0.002957
1989    0.003769
1990    0.008226
1991    0.007413
1992    0.008142
1993    0.009267
1994    0.013744
1995    0.027321
1996    0.030070
1997    0.042232
1998    0.051082
1999    0.062452
2000    0.069844
2001    0.056267
2002    0.052748
2003    0.056788
2004    0.056996
2005    0.062785
2006    0.056371
2007    0.047979
2008    0.046459
2009    0.043689
2010    0.033256
2011    0.034027
2012    0.027550
2013    0.016784
2014    0.0138

Now that the few incorrect values are removed, we can see the accurate data above. Registering vehicles listed on eBay started gaining momentum in the 90's, based on these numbers. The majority of the vehicles in our dataset have been registered in approximately the last 20 years.

# Brand

In [21]:
autos["brand"].value_counts(normalize=True)

volkswagen        0.212136
bmw               0.110035
opel              0.108161
mercedes_benz     0.095354
audi              0.086400
ford              0.069782
renault           0.047354
peugeot           0.029529
fiat              0.025864
seat              0.018180
skoda             0.016035
mazda             0.015139
nissan            0.015098
citroen           0.013911
smart             0.013911
toyota            0.012474
sonstige_autos    0.010912
hyundai           0.009850
volvo             0.009246
mini              0.008642
mitsubishi        0.008142
honda             0.007851
kia               0.007101
alfa_romeo        0.006622
porsche           0.006101
suzuki            0.005914
chevrolet         0.005706
chrysler          0.003665
daihatsu          0.002561
dacia             0.002561
jeep              0.002249
subaru            0.002187
land_rover        0.002041
saab              0.001603
jaguar            0.001583
trabant           0.001562
daewoo            0.001499
r

Several brands hold a large number of the data: Volkswagen, BMW, and Opel. Others hold a minescule amount. To reduce the amount of brands we are going to aggregate by, we'll drop all brands that hold less than 5% of the total values. 

In [22]:
brand_values = autos["brand"].value_counts(normalize=True)
chosen_brands = brand_values[brand_values > .05].index
print(chosen_brands)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')


In [23]:
brand_mean_prices = {}

for brand in chosen_brands:
    brand_only = autos[autos["brand"] == brand]
    mean_price = brand_only["price"].mean()
    brand_mean_prices[brand] = int(mean_price)
    
brand_mean_prices

{'volkswagen': 5426,
 'bmw': 8334,
 'opel': 2876,
 'mercedes_benz': 8485,
 'audi': 9093,
 'ford': 3949}

The dictionary above shows the brand along with it's mean price. Audi, Mercedes Benz, and BMW have the highest prices by far. Volkswagen in in the mid range, and  For and Opel are at the lower end of price.

# Mileage

In [24]:
bmp_series = pd.Series(brand_mean_prices)
print(bmp_series)

volkswagen       5426
bmw              8334
opel             2876
mercedes_benz    8485
audi             9093
ford             3949
dtype: int64


In [28]:
brand_mean_mileage = {}

for brand in chosen_brands:
    brand_only = autos[autos["brand"] == brand]
    mean_mileage = brand_only["odometer_km"].mean()
    brand_mean_mileage[brand] = int(mean_mileage)

mean_mileage = pd.Series(brand_mean_mileage).sort_values(ascending=False)
mean_prices = pd.Series(brand_mean_prices).sort_values(ascending=False)

brand_info = pd.DataFrame(mean_mileage,columns=['mean_mileage'])
brand_info


mean_mileage
bmw                  132434
mercedes_benz        130856
audi                 129287
opel                 129223
volkswagen           128728
ford                 124068

In [29]:
brand_info["mean_price"] = mean_prices
brand_info

mean_mileage  mean_price
bmw                  132434        8334
mercedes_benz        130856        8485
audi                 129287        9093
opel                 129223        2876
volkswagen           128728        5426
ford                 124068        3949

When seen side by side in the chart above, we can tell that the price difference by brand does not have any correlation with mean mileage. In fact, the higher price vehicles have a slightly higher mean mileage than the lower priced vehicles. 